sympy/core/basic.py

```diff
@@ -1228,31 +1229,88 @@ def has(self, *patterns):
         False
 
         """
-        return any(self._has(pattern) for pattern in patterns)
+        return self._has(iterargs, *patterns)
 
-    def _has(self, pattern):
-        """Helper for .has()"""
-        from .function import UndefinedFunction, Function
-        if isinstance(pattern, UndefinedFunction):
-            return any(pattern in (f, f.func)
-                       for f in self.atoms(Function, UndefinedFunction))
+    @cacheit
+    def has_free(self, *patterns):
+        """return True if self has object(s) ``x`` as a free expression
+        else False.
 
-        if isinstance(pattern, BasicMeta):
-            subtrees = _preorder_traversal(self)
-            return any(isinstance(arg, pattern) for arg in subtrees)
+        Examples
+        ========
 
-        pattern = _sympify(pattern)
+        >>> from sympy import Integral, Function
+        >>> from sympy.abc import x, y
+        >>> f = Function('f')
+        >>> g = Function('g')
+        >>> expr = Integral(f(x), (f(x), 1, g(y)))
+        >>> expr.free_symbols
+        {y}
+        >>> expr.has_free(g(y))
+        True
+        >>> expr.has_free(*(x, f(x)))
+        False
 
-        _has_matcher = getattr(pattern, '_has_matcher', None)
-        if _has_matcher is not None:
-            match = _has_matcher()
-            return any(match(arg) for arg in _preorder_traversal(self))
-        else:
-            return any(arg == pattern for arg in _preorder_traversal(self))
+        This works for subexpressions and types, too:
+
+        >>> expr.has_free(g)
+        True
+        >>> (x + y + 1).has_free(y + 1)
+        True
+
+        """
+        return self._has(iterfreeargs, *patterns)
+
+    def _has(self, iterargs, *patterns):
+        # separate out types and unhashable objects
+        type_set = set()  # only types
+        p_set = set()  # hashable non-types
+        other = []
+        for p in patterns:
+            if isinstance(p, BasicMeta):
+                type_set.add(p)
+                continue
+            if not isinstance(p, Basic):
+                try:
+                    p = _sympify(p)
+                except SympifyError:
+                    continue  # Basic won't have this in it
+            try:
+                p_set.add(p)
+            except TypeError:
+                other.append(p)
+
+        for i in iterargs(self):
+            try:
+                if i in p_set:
+                    return True
+            except TypeError:
+                pass  # unhashable i, e.g. PermutationGroup([(0 2 1)])
```

Comment:
Are there other unhashable Basic types besides PermutationGroup?

Reply:
This is the only one that triggered. Why is it not hashable? It derives from Basic.

Reply:
A subclass that refines `__eq__` must also redefine `__hash__` to be hashable. The `__hash__` method for `PermutationGroup` would probably be quite expensive if it should match the behaviour of `__eq__` (`x == y  -->  hash(x)==hash(y)`). Really it is best not to redefine `__eq__` in `Basic` subclasses and to accept that `a == b` is for structural equality.